### ISSUE TITLE
Overlays and Statement Generator do not appear to work together

### ISSUE THREAD

**Issue by lukegalea** [ulfurinn/wongi-engine#62]
See https://github.com/ulfurinn/wongi-engine/pull/61

No idea why, but when we use overlays with a statement generator we are finding that the token doesn't include an overlay.

This fix resolves the problem, but I imagine it'll have some terrible consequence in the future that we haven't noticed yet...

We are invoking the statement generator from within an action like this:

```
      fact = Wongi::Engine::Template.new(membership, key, value)

      generator = Wongi::Engine::DSL::Action::StatementGenerator.new fact
      generator.rete = rete
      generator.production = production
      generator.execute token
```

**Comment by lukegalea** [comment]
I realize now there are bunch of other places that need similar guards. I think a better solution is to figure out why token.overlay is nil...

**Comment by ulfurinn** [comment]
Why are using the generator directly? It's not part of the public API.

**Comment by lukegalea** [comment]
@ulfurinn We have some actions that need to generate new facts afterwards. Specific example is an action the enrolls a client in a course, then using the statement generator to define facts about the newly created enrollment.

**Comment by ulfurinn** [comment]
Why can't you use the regular `gen` clause?

Have you confirmed that is it the manual use of the generator that's causing whatever problem you're observing?

**Comment by lukegalea** [comment]
@ulfurinn I'm not sure how I can use the gen clause within an action (inside the "#run" method). The deal is that we have an action that creates a new model and needs to define a fact about that new model.  

I haven't confirmed that there are any issues with the gen clause, but only because it's not clear how I could use it :)

**Comment by ulfurinn** [comment]
> I'm not sure how I can use the gen clause within an action (inside the "#run" method).

You can't and you're not supposed to. You put it after the custom action. Custom actions are for side-effects **only**.

If you can calculate the elements of the new facts before the action, put them in `assign` blocks, otherwise I'll need to add the ability to add arbitrary data to tokens to propagate them between steps (although I'd generally recommend restructuring the rules to put all side-effects last, adding surrogate facts if necessary. If you're putting domain objects in the dataset, consider having a translation layer instead. I cannot give more advice without knowing what you're doing).

> I haven't confirmed that there are any issues with the gen clause, but only because it's not clear how I could use it :)

Ignore your custom actions and add any random `gen` clause to make sure the overlay gets linked correctly.

**Comment by lukegalea** [comment]
@ulfurinn thanks for your help. The side effect inside the action actually needs to be part of the new fact that's generated.

There's a couple places we use the statement generator. One is where a special subset of the rules need detailed (expensive) analytics to be calculated.  

```
rule "L2 rules require membership stats" do
    forall {
      has :Event, 'program_name', 'Level 2 Coaching Certification'
    }

    do! {
      calculate_membership_stats
    }
  end

  rule "Course passed" do
    forall {
      assuming 'L2 course event'

      greater :AverageScore, 75
      greater :LessonCompletionRate, 0.75
      greater :HabitCompletionRate, 0.75
    }
```

calculate_membership_stats above defines :AverageScore, :LessonCompletionRate and :HabitCompletionRate

We could work around it by deciding if we need to calculate these stats and define the facts in ruby instead of in the rules engine, but I really like how it works now.

**Comment by lukegalea** [comment]
@ulfurinn We'd be happy to sponsor your work on this. We definitely need some way to have an action generate facts, via whatever syntax you think is best.

**Comment by ulfurinn** [comment]
We've discussed this already.

I'll add a facility to propagate extra data through the token.

As for the original issue, please post a full example to reproduce, or I will close the ticket.

**Comment by lukegalea** [comment]
Thanks @ulfurinn. Just trying to help support your work.

Propogating data through the token sounds great. I think it's fair to close this issue because, as you said, it's caused by us abusing an internal API.

**Comment by ulfurinn** [comment]
This is available in 0.3.0:

https://github.com/ulfurinn/wongi-engine/commit/afa273e141841ee374fa5bf0920b13f31bb0ffc0#diff-371c522f25b007893689415bb94c6c1f
